Migrating basic Multiplayer testing workflows (#80)
o3de/sig-network: testing-guidance/workflow-tests/multiplayer/basic-multiplayer-workflow-tests.md

# Multiplayer Testing

These tests are designed to cover the basic functionality related to running a level from a project that is intended to be multiplayer.

> General Documentation for this feature set: [https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/](https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/)

Reviewer: ```suggestion
General Documentation for this feature set: [https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/](https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/).
```

## Prerequisites

*   MultiplayerSample project is cloned and built ( [https://github.com/o3de/o3de-multiplayersample](https://github.com/o3de/o3de-multiplayersample) )

>     *   If using this approach, the SampleBase level can be used for most tests

Reviewer: ```suggestion
    * If using this approach, the SampleBase level can be used for most tests.
    * Ensure the ServerLauncher is built, as well as the Editor and ClientLauncher (if testing via the client launcher).
```

*   Alternatively: A project has been created and built that includes the Multiplayer Gem and subsequent changes have been made to enable functionality ( [https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/configuration/](https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/configuration/) )

>     *   If using this approach, a level should be created in the project that utilizes a player spawner and network prefabs, as these will be easy identifiers for whether multiplayer connections are working as intended

Reviewer: ```suggestion
    *   If using this approach, a new level should be created in the project, The level should utilize a player spawner and network prefabs, as these will act as identifiers for whether multiplayer connections are working as intended.
```

## Common Issues

*   Most issues will be caught through log errors, which can be seen in the Editor and Launcher consoles or in logs after running tests
*   In addition, network prefabs can be a visual clue that something isn't working as expected (objects disappearing or not appearing in the first place)

> *   Processes that remain after closing applications - this can result in errors when trying to launch subsequent times

Reviewer: ```suggestion
* Issues connecting to the server are caused either by:
       1. Processes that remain after closing applications - this can result in errors when trying to launch subsequent times.
       2. The ServerLauncher has not been built or has not been built against the same commit as Editor/ClientLauncher.
```

> **Multiplayer** **Testing**

Reviewer: IMHO we should Follow o3de.org formatting rules for documentation so use ## for headings rather than underlines etc https://www.o3de.org/docs/contributing/to-docs/style-guide/quick-reference/

===========================

These tests are designed to cover the basic functionality related to running a level from a project that is intended to be multiplayer.

General Documentation for this feature set: [https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/](https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/)

**Prerequisites**
=================

> *   MultiplayerSample project is acquired and built ( [https://github.com/o3de/o3de-multiplayersample](https://github.com/o3de/o3de-multiplayersample) )

Reviewer: nit: 
```suggestion
*   MultiplayerSample project is cloned and built ( [https://github.com/o3de/o3de-multiplayersample](https://github.com/o3de/o3de-multiplayersample) )
```

    *   If using this approach, the SampleBase level can be used for most tests

> *   Alternatively: A project has been created and built that includes the Multiplayer Gem and subsequent changes have been made to enable functionality ( [https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/configuration/](https://docs.o3de.org/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/configuration/) )

Reviewer: Do we want to cover the new Multiplayer Template here?

Author: I'll do an update later when the Multiplayer Template is officially ready for download

>     *   If using this approach, a level should be created in the project that utilizes a spawner and network prefabs, as these will be easy identifiers for whether multiplayer connections are working as intended

Reviewer: nit: player spawner? We want to at some point point to Gene's O3DCon workshop talk for a simple tutorial to setting up multiplayer content.

**Common Issues**
=================

*   Most issues will be caught through log errors

> *   In addition, network prefabs can be a visual clue that something isn't working as expected

Reviewer: Talk about debug text when playing in error.

Maybe talk about content disappearing when playing in game, indicates that content is not correctly networked? Ideally we would link to the standard troubleshooting guide for networking, but that doesn;t exist.

**Workflows**
=============

  

> | Workflow                        | Requests                                                                                                                                                                                                                                                                                                                                                    | What To Watch Out For                                                                                                                                                                   |

Reviewer: Maybe cover connection errors caused by zombie servers? or client/server version mismatch?
